testutil/ciao-down/cloudinit.go

```diff
@@ -0,0 +1,427 @@
+  "hostname": "singlevm"
+}
+`
+
+// TODO: Code copied from launcher.  Needs to be moved to qemu
+
+func createCloudInitISO(ctx context.Context, instanceDir string, userData, metaData []byte) error {
+	configDrivePath := path.Join(instanceDir, "clr-cloud-init")
+	dataDirPath := path.Join(configDrivePath, "openstack", "latest")
+	metaDataPath := path.Join(dataDirPath, "meta_data.json")
+	userDataPath := path.Join(dataDirPath, "user_data")
+	isoPath := path.Join(instanceDir, "config.iso")
+
+	defer func() {
+		_ = os.RemoveAll(configDrivePath)
+	}()
+
+	err := os.MkdirAll(dataDirPath, 0755)
+	if err != nil {
+		glog.Errorf("Unable to create config drive directory %s", dataDirPath)
+		return err
+	}
+
+	err = ioutil.WriteFile(metaDataPath, metaData, 0644)
+	if err != nil {
+		glog.Errorf("Unable to create %s", metaDataPath)
+		return err
+	}
+
+	err = ioutil.WriteFile(userDataPath, userData, 0644)
+	if err != nil {
+		glog.Errorf("Unable to create %s", userDataPath)
+		return err
+	}
+
+	cmd := exec.CommandContext(ctx, "xorriso", "-as", "mkisofs", "-R", "-V", "config-2",
+		"-o", isoPath, configDrivePath)
+	err = cmd.Run()
+	if err != nil {
+		return fmt.Errorf("Unable to create cloudinit iso image %v", err)
+	}
+
+	return nil
+}
+
+func buildISOImage(ctx context.Context, instanceDir string, ws *workspace) error {
+	udt := template.Must(template.New("user-data").Parse(userDataTemplate))
+	var udBuf bytes.Buffer
+	err := udt.Execute(&udBuf, ws)
+	if err != nil {
+		return fmt.Errorf("Unable to execute user data template : %v", err)
+	}
+
+	mdt := template.Must(template.New("meta-data").Parse(metaDataTemplate))
+
+	var mdBuf bytes.Buffer
+	err = mdt.Execute(&mdBuf, ws)
+	if err != nil {
+		return fmt.Errorf("Unable to execute user data template : %v", err)
+	}
+
+	return createCloudInitISO(ctx, instanceDir, udBuf.Bytes(), mdBuf.Bytes())
+}
+
+// TODO: Code copied from launcher.  Needs to be moved to qemu
+
+func createRootfs(ctx context.Context, backingImage, instanceDir string) error {
+	vmImage := path.Join(instanceDir, "image.qcow2")
+	if _, err := os.Stat(vmImage); err == nil {
+		_ = os.Remove(vmImage)
+	}
+	params := make([]string, 0, 32)
+	params = append(params, "create", "-f", "qcow2", "-o", "backing_file="+backingImage,
+		vmImage, "60000M")
+	return exec.CommandContext(ctx, "qemu-img", params...).Run()
+}
+
+func bootVM(ctx context.Context, ws *workspace) error {
+	disconnectedCh := make(chan struct{})
+	socket := path.Join(ws.instanceDir, "socket")
+	qmp, _, err := qemu.QMPStart(ctx, socket, qemu.QMPConfig{}, disconnectedCh)
+	if err == nil {
+		qmp.Shutdown()
+		return fmt.Errorf("VM is already running")
+	}
+
+	vmImage := path.Join(ws.instanceDir, "image.qcow2")
+	isoPath := path.Join(ws.instanceDir, "config.iso")
+	fsdevParam := fmt.Sprintf("local,security_model=passthrough,id=fsdev0,path=%s",
+		ws.GoPath)
+	args := []string{
+		"-qmp", fmt.Sprintf("unix:%s,server,nowait", socket), "-m", "8G", "-smp", "cpus=2",
```

Comment:
Can the 8G and 2 vcpu's be config knobs?  I'm liable to want to set those higher.  The cpu's I'd personally do higher be default.  The RAM I'd sometimes make more on my beefy dev box, but might also prefer less.  8G's a lot if you're running on an ultramobile laptop doing dev on the go and only have 8 or 16G total system RAM.  This also needs documented with respect to how it will interrelate with how many guest workload instances you can launch within the single VM an their cumulative resource consumption.